### PR TITLE
[ALBEF] Text Encoder

### PR DESCRIPTION
**Pull request by katrina433**
Summary:

This PR implements the text encoder used in the ALBEF model. 

The text encoder is based off of the [bert implementation](https://github.com/salesforce/ALBEF/blob/6224e78e5292757ca9c7d5add3e415010955a58b/models/xbert.py) in the ALBEF repo. Only classes and operations that affect the final output are kept in this PR. 

TODO: All of the text encoder's components are in the same file as the text encoder, but will refactor once the multimodal encoder is finished. 

Test plan:

Initiated the text encoder and loaded the pre-trained bert `state_dict` from the ALBEF repo, and checked that the outputs match. 

Wrote unit tests with small inputs and invalid inputs to check for expected outputs/errors.

### CODE REVIEW

File: test/modules/encoders/test_albef_text_encoder.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            [
+                [[-0.668618, -0.744909, 1.413527], [-0.643172, 1.412341, -0.769169]],
+                [[-1.052131, -0.292326, 1.344457], [0.986411, 0.384430, -1.370841]],
+            ]
+        )
+        assert_expected(output, expected, rtol=0, atol=1e-4)
+
+    def test_text_encoder_without_attention_mask(self):
+        set_rng_seed(0)
+        input_ids = torch.randint(10, (2, 2))
```
Comment (ebsmothers) [inline test/modules/encoders/test_albef_text_encoder.py:33]: So input_ids will be the same in the first two test cases but not the last two? Wondering why you set the seed in some, but not all, test cases

File: torchmultimodal/modules/encoders/albef_text_encoder.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        # Normalize the attention scores to probabilities.
+        attention_probs = nn.Softmax(dim=-1)(attention_scores)
+
+        context_layer = torch.matmul(attention_probs, value_layer)
+
+        context_layer = context_layer.permute(0, 2, 1, 3).contiguous()
+        new_context_layer_shape = context_layer.size()[:-2] + (self.all_head_size,)
+        context_layer = context_layer.view(*new_context_layer_shape)
+
+        return context_layer
```
Comment (ebsmothers) [inline torchmultimodal/modules/encoders/albef_text_encoder.py:284]: I see a lot of similarities between these classes and the corresponding [FLAVA ones](https://github.com/facebookresearch/multimodal/blob/a3206ebab51157fd78ffcaee111045d725cf8c6c/torchmultimodal/modules/layers/transformer.py#L26-L257). Not essential for you to generalize in this PR, but definitely something we will want to do later on. cc @RdoubleA

File: torchmultimodal/modules/encoders/albef_text_encoder.py [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+        inputs_embeds = self.word_embeddings(input_ids)
+        position_embeddings = self.position_embeddings(position_ids)
+        token_type_embeddings = self.token_type_embeddings(token_type_ids)
+
+        embeddings = inputs_embeds + position_embeddings + token_type_embeddings
+        embeddings = self.LayerNorm(embeddings)
+        return embeddings
+
+
+class ALBEFEncoder(nn.Module):
```
Comment (ebsmothers) [inline torchmultimodal/modules/encoders/albef_text_encoder.py:148]: Can we name this something else? `ALBEFEncoder` and `ALBEFTextEncoder` in the same file (not to mention other `ALBEF*Encoders` we have floating around) can be confusing

File: torchmultimodal/modules/encoders/albef_text_encoder.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+                for _ in range(num_hidden_layers)
+            ]
+        )
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.FloatTensor] = None,
+    ) -> Tensor:
+        for _, layer_module in enumerate(self.layer):
```
Comment (ebsmothers) [inline torchmultimodal/modules/encoders/albef_text_encoder.py:175]: Why enumerate here?

File: test/modules/encoders/test_albef_text_encoder.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+
+class TestALBEFTextEncoder:
+    set_rng_seed(0)
+    text_encoder = ALBEFTextEncoder(hidden_size=3, num_attention_heads=1)
+
+    def test_text_encoder(self):
+        set_rng_seed(0)
+        input_ids = torch.randint(10, (2, 2))
+        text_atts = torch.randn(2, 2)
```
Comment (ebsmothers) [inline test/modules/encoders/test_albef_text_encoder.py:21]: Attentions should be in {0,1}, no?

File: test/modules/encoders/test_albef_text_encoder.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import pytest
+import torch
+from test.test_utils import assert_expected, set_rng_seed
+from torch import Tensor
+from torchmultimodal.modules.encoders.albef_text_encoder import ALBEFTextEncoder
+
+
+class TestALBEFTextEncoder:
+    set_rng_seed(0)
+    text_encoder = ALBEFTextEncoder(hidden_size=3, num_attention_heads=1)
```
[the remark this answers is not in the document]
Comment (katrina433) [reply test/modules/encoders/test_albef_text_encoder.py:16]: The reason I had `set_rng_seed` in every method is because if I only set seed once in `init`, the outputs are different when I run `pytest` on a single file and run `pytest` on all test files. I just tried fixture and it seems to be the same case.

File: test/modules/encoders/test_albef_text_encoder.py (hunk 1 above)
Comment (katrina433) [reply test/modules/encoders/test_albef_text_encoder.py:33]: The last two test cases I passed in invalid inputs to test for errors raised, so the exact input values aren't really relevant.

File: torchmultimodal/modules/encoders/albef_text_encoder.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+        )
+        self.encoder = ALBEFEncoder(
+            hidden_size,
+            intermediate_size,
+            num_attention_heads,
+            num_hidden_layers,
+            layer_norm_eps,
+        )
+
+    def get_extended_attention_mask(self, attention_mask: Tensor) -> Tensor:
```
[the remark this answers is not in the document]
Comment (katrina433) [reply torchmultimodal/modules/encoders/albef_text_encoder.py:66]: Should I also remove this method in FLAVA and import from utils?

File: test/modules/encoders/test_albef_text_encoder.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (langong347) [reply test/modules/encoders/test_albef_text_encoder.py:16]: The fixture should be called/refreshed by each test method. 

How about this:
```
@pytest.fixture(autouse=True)
def some_fixture(self):
    set_rng_seed(40)
```
By setting `autouse=True` the fixture will be automatically called in each of the test method.

[the remark this answers is not in the document]
Comment (ebsmothers) [reply test/modules/encoders/test_albef_text_encoder.py:16]: @katrina433 can you follow @langong347's suggestion here?

**Review by ebsmothers — APPROVED** [review]
Modulo a couple minor comments on the test, this looks good to me

File: torchmultimodal/modules/encoders/albef_text_encoder.py [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+        padding_idx: int,
+        max_position_embeddings: int,
+        type_vocab_size: int,
+        layer_norm_eps: float,
+    ) -> None:
+        super().__init__()
+        self.word_embeddings = nn.Embedding(vocab_size, hidden_size, padding_idx)
+        self.position_embeddings = nn.Embedding(max_position_embeddings, hidden_size)
+        self.token_type_embeddings = nn.Embedding(type_vocab_size, hidden_size)
+        self.LayerNorm = nn.LayerNorm(hidden_size, eps=layer_norm_eps)
```
Comment (RdoubleA) [inline torchmultimodal/modules/encoders/albef_text_encoder.py:99]: I would change the name of this slighty so it's not too similar to `nn.LayerNorm`
```suggestion
        self.layer_norm = nn.LayerNorm(hidden_size, eps=layer_norm_eps)
```

File: torchmultimodal/modules/encoders/albef_text_encoder.py [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+                    layer_norm_eps,
+                )
+                for _ in range(num_hidden_layers)
+            ]
+        )
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.FloatTensor] = None,
```
Comment (RdoubleA) [inline torchmultimodal/modules/encoders/albef_text_encoder.py:144]: just make this normal tensor like other `attention_mask`s

File: torchmultimodal/modules/encoders/albef_text_encoder.py [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+                    num_attention_heads,
+                    layer_norm_eps,
+                )
+                for _ in range(num_hidden_layers)
+            ]
+        )
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
```
Comment (RdoubleA) [inline torchmultimodal/modules/encoders/albef_text_encoder.py:143]:
```suggestion
        hidden_states: Tensor,
```

File: torchmultimodal/modules/encoders/albef_text_encoder.py [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+        hidden_size: int,
+        num_attention_heads: int,
+        layer_norm_eps: float,
+    ) -> None:
+        super().__init__()
+        self.self_attention = ALBEFTransformerSelfAttention(
+            hidden_size, num_attention_heads
+        )
+        self.dense = nn.Linear(hidden_size, hidden_size)
+        self.LayerNorm = torch.nn.LayerNorm(hidden_size, eps=layer_norm_eps)
```
Comment (RdoubleA) [inline torchmultimodal/modules/encoders/albef_text_encoder.py:193]:
```suggestion
        self.LayerNorm = nn.LayerNorm(hidden_size, eps=layer_norm_eps)
```

File: torchmultimodal/utils/common.py [hunk 12]
```diff
@@ -21,6 +21,37 @@ def get_current_device():
         return torch.device("cpu")
 
 
+def get_extended_attention_mask(attention_mask: Tensor) -> Tensor:
```
Comment (RdoubleA) [inline torchmultimodal/utils/common.py:24]: I am not sure I would put this in utils/common, it's too specific to attention. I think for this PR it's fine to leave it here but we should consider a better location for this, such as layers/attention or utils/attention

File: torchmultimodal/modules/encoders/albef_text_encoder.py [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+
+        self.num_attention_heads = num_attention_heads
+        self.attention_head_size = int(hidden_size / num_attention_heads)
+        self.all_head_size = self.num_attention_heads * self.attention_head_size
+
+        self.query = nn.Linear(hidden_size, self.all_head_size)
+        self.key = nn.Linear(hidden_size, self.all_head_size)
+        self.value = nn.Linear(hidden_size, self.all_head_size)
+
+    def transpose_for_scores(self, x: Tensor) -> Tensor:
```
Comment (RdoubleA) [inline torchmultimodal/modules/encoders/albef_text_encoder.py:227]: can we refactor this and the FLAVA `transpose_for_scores` into a utility so we don't have two copies of this?

File: torchmultimodal/modules/encoders/albef_text_encoder.py [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+        new_x_shape = x.size()[:-1] + (
+            self.num_attention_heads,
+            self.attention_head_size,
+        )
+        x = x.view(*new_x_shape)
```
Comment (RdoubleA) [inline torchmultimodal/modules/encoders/albef_text_encoder.py:232]: you could simplify these lines with `unflatten`, i.e.:
```
x = x.unflatten(-1, (self.num_attention_heads, self.attention_head_size))
```

File: torchmultimodal/modules/encoders/albef_text_encoder.py [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+    Args:
+        vocab_size (int): Vocabulary size of the model. Defines the different tokens that can be represented by the inputs_ids.
+        hidden_size (int): Dimensionality of the encoder layers.
+        num_hidden_layers (int): Number of hidden layers in the Transformer encoder.
+        num_attention_heads (int): Number of attention heads for each attention layer in the Transformer encoder.
+        intermediate_size (int): Dimensionality of the “intermediate” (i.e., feed-forward) layer in the Transformer encoder.
+        max_position_embeddings (int): The maximum sequence length that this model might ever be used with.
+        type_vocab_size (int): The vocabulary size of the token_type_ids.
+        pad_token_id (int): The embedding for pad_token_id is not updated during training.
+        layer_norm_eps (float): The epsilon used by the layer normalization layers.
```
Comment (RdoubleA) [inline torchmultimodal/modules/encoders/albef_text_encoder.py:30]: nit: add `Default is ___.` to the ends of these lines since there are default values. docstring convention :)

File: torchmultimodal/modules/encoders/albef_text_encoder.py (hunk 2 above)
Comment (RdoubleA) [reply torchmultimodal/modules/encoders/albef_text_encoder.py]: Yeah, I think this is the third or fourth implementation of attention and transformer, so will definitely need to make a note of this

**Review by RdoubleA — APPROVED** [review]
Thanks for doing this! Let's also make a note that FLAVA needs to import `transpose_for_scores` from utils. Maybe I'll make a task
